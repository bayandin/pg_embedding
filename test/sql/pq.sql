CREATE TABLE t (id integer primary key, val real[]);
select setseed(0.5);
INSERT INTO t select id, array[random(), random(), random(), random()] from generate_series(1,10000) id;
CREATE INDEX hnsw_pq_idx ON t USING hnsw (val) WITH (dims=4, m=8, efConstruction=16, efSearch=10, pqBits=3, pqSubqs=2);
CREATE INDEX ON t USING hnsw (val) WITH (dims=4, m=8, efConstruction=16, efSearch=10);

SET enable_seqscan = off;
explain SELECT * FROM t ORDER BY val <-> array[0.5,0.5,0.5,0.5] limit 10;
SELECT * FROM t ORDER BY val <-> array[0.5,0.5,0.5,0.5] limit 10;

SET enable_seqscan = on;
SET enable_indexscan = off;
explain SELECT * FROM t ORDER BY val <-> array[0.5,0.5,0.5,0.5] limit 10;
SELECT * FROM t ORDER BY val <-> array[0.5,0.5,0.5,0.5] limit 10;
select pg_relation_size('hnsw_pq_idx');

drop index hnsw_pq_idx;
CREATE INDEX hnsw_idx ON t USING hnsw (val) WITH (dims=4, m=8, efConstruction=16, efSearch=10);
SELECT * FROM t ORDER BY val <-> array[0.5,0.5,0.5,0.5] limit 10;
select pg_relation_size('hnsw_idx');

DROP TABLE t;
