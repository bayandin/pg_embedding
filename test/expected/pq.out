CREATE TABLE t (id integer primary key, val real[]);
select setseed(0.5);
 setseed 
---------
 
(1 row)

INSERT INTO t select id, array[random(), random(), random(), random()] from generate_series(1,10000) id;
CREATE INDEX hnsw_pq_idx ON t USING hnsw (val) WITH (dims=4, m=8, efConstruction=16, efSearch=10, pqBits=3, pqSubqs=2);
CREATE INDEX ON t USING hnsw (val) WITH (dims=4, m=8, efConstruction=16, efSearch=10);
SET enable_seqscan = off;
explain SELECT * FROM t ORDER BY val <-> array[0.5,0.5,0.5,0.5] limit 10;
                                   QUERY PLAN                                    
---------------------------------------------------------------------------------
 Limit  (cost=40.00..40.07 rows=10 width=40)
   ->  Index Scan using t_val_idx on t  (cost=40.00..105.10 rows=10000 width=40)
         Order By: (val <-> '{0.5,0.5,0.5,0.5}'::real[])
(3 rows)

SELECT * FROM t ORDER BY val <-> array[0.5,0.5,0.5,0.5] limit 10;
  id  |                      val                      
------+-----------------------------------------------
  400 | {0.46710503,0.4545811,0.45946807,0.5224157}
 9221 | {0.5693331,0.4639949,0.5038537,0.4708719}
 1678 | {0.4302664,0.46579468,0.5254703,0.44620183}
 2272 | {0.45853564,0.55112475,0.57308996,0.49787247}
 7753 | {0.46425086,0.45226237,0.41756374,0.5148064}
 6613 | {0.5985818,0.5113737,0.47121355,0.51103425}
 2302 | {0.5201298,0.49001274,0.5395254,0.59521574}
 5407 | {0.5559473,0.42566732,0.54485106,0.47261426}
 4674 | {0.48284096,0.39664084,0.5088823,0.47825432}
 6795 | {0.42835304,0.5504437,0.5303207,0.5773013}
(10 rows)

SET enable_seqscan = on;
SET enable_indexscan = off;
explain SELECT * FROM t ORDER BY val <-> array[0.5,0.5,0.5,0.5] limit 10;
                             QUERY PLAN                             
--------------------------------------------------------------------
 Limit  (cost=435.10..435.12 rows=10 width=40)
   ->  Sort  (cost=435.10..460.10 rows=10000 width=40)
         Sort Key: ((val <-> '{0.5,0.5,0.5,0.5}'::real[]))
         ->  Seq Scan on t  (cost=0.00..219.00 rows=10000 width=40)
(4 rows)

SELECT * FROM t ORDER BY val <-> array[0.5,0.5,0.5,0.5] limit 10;
  id  |                      val                      
------+-----------------------------------------------
  400 | {0.46710503,0.4545811,0.45946807,0.5224157}
 9221 | {0.5693331,0.4639949,0.5038537,0.4708719}
 1678 | {0.4302664,0.46579468,0.5254703,0.44620183}
 2272 | {0.45853564,0.55112475,0.57308996,0.49787247}
 7753 | {0.46425086,0.45226237,0.41756374,0.5148064}
 6613 | {0.5985818,0.5113737,0.47121355,0.51103425}
 2302 | {0.5201298,0.49001274,0.5395254,0.59521574}
 5407 | {0.5559473,0.42566732,0.54485106,0.47261426}
 4674 | {0.48284096,0.39664084,0.5088823,0.47825432}
 6795 | {0.42835304,0.5504437,0.5303207,0.5773013}
(10 rows)

select pg_relation_size('hnsw_pq_idx');
 pg_relation_size 
------------------
           540672
(1 row)

drop index hnsw_pq_idx;
CREATE INDEX hnsw_idx ON t USING hnsw (val) WITH (dims=4, m=8, efConstruction=16, efSearch=10);
SELECT * FROM t ORDER BY val <-> array[0.5,0.5,0.5,0.5] limit 10;
  id  |                      val                      
------+-----------------------------------------------
  400 | {0.46710503,0.4545811,0.45946807,0.5224157}
 9221 | {0.5693331,0.4639949,0.5038537,0.4708719}
 1678 | {0.4302664,0.46579468,0.5254703,0.44620183}
 2272 | {0.45853564,0.55112475,0.57308996,0.49787247}
 7753 | {0.46425086,0.45226237,0.41756374,0.5148064}
 6613 | {0.5985818,0.5113737,0.47121355,0.51103425}
 2302 | {0.5201298,0.49001274,0.5395254,0.59521574}
 5407 | {0.5559473,0.42566732,0.54485106,0.47261426}
 4674 | {0.48284096,0.39664084,0.5088823,0.47825432}
 6795 | {0.42835304,0.5504437,0.5303207,0.5773013}
(10 rows)

select pg_relation_size('hnsw_idx');
 pg_relation_size 
------------------
           688128
(1 row)

DROP TABLE t;
